### ISSUE TITLE
illogically high VRAM usage for `1.5B` model

### ISSUE THREAD

**Issue by skoshx** [willccbb/verifiers#26]
Here's a screenshot of `nvidia-smi` when running with the smallest 1.5B model calculator example.

![Image](https://github.com/user-attachments/assets/e1591d48-9e3b-42c1-9bcf-6eb2681e6161)

It makes no sense to me how it can be so high? Even with the reference model the weights should be only `2 (trained model + reference model) * 1.5B (model params) * 2 (bytes per param - Bfloat16 precision) = 6GB`, and throw in optimizer states etc, should be nowhere near 44GB of VRAM?

Am I missing something?

**Comment by willccbb** [comment]
The cuda:1 VRAM is just from the vLLM memory limit, it'll use 0.7 by default. Then 6GB for model copies, 12GB for optimizer states, and then more for activations (which add up quickly when you have lots of rollouts per minibatch, needs to backprop through all logits at each token). If you set bs/rollouts to be small + truncate prompts/completions to be super short (32-64 toks) it should initially be closer to 20gb.

Though, you shouldn't worry much about the reported memory number if it's still far from the max available -- the deepspeed/accelerate setup isn't the most aggressive at clearing out the cache, so your reported memory usage will be higher than the amount actually needed to run. The cache is cleared more aggressively once you're closer to headroom limits. "Keep increasing size until you actually hit OOM" is the most reliable strategy to measure actual requirements here in my experience (unfortunately).
